dplython/dplython.py

```diff
@@ -473,23 +483,78 @@ def mutating_join(right, **kwargs):
     dsuffixes = kwargs['suffixes']
   else:
     dsuffixes = ('_x', '_y')
-  def f(df):
-    x = lambda df: DplyFrame(df.merge(right, how=kwargs['how'], left_on=left_cols, right_on=right_cols, suffixes=dsuffixes))
-    return x
-  return f
+  if left._grouped_on:
+    outDf = (DplyFrame((left >> ungroup())
+                       .merge(right, how=kwargs['how'], left_on=left_cols, right_on=right_cols, suffixes=dsuffixes))
+             .regroup(left._grouped_on))
+  else:
+    outDf = DplyFrame(left.merge(right, how=kwargs['how'], left_on=left_cols, right_on=right_cols, suffixes=dsuffixes))
+  return outDf
 
-@ApplyToDataframe
-def inner_join(right, **kwargs):
-  return mutating_join(right, how='inner', **kwargs)
 
-@ApplyToDataframe
-def full_join(right, **kwargs):
-  return mutating_join(right, how='outer', **kwargs)
+class MutatingJoin(Verb):
 
-@ApplyToDataframe
-def left_join(right, **kwargs):
-  return mutating_join(right, how='left', **kwargs)
+  def __new__(cls, *args, **kwargs):
+    if len(args) > 1 and isinstance(args[0], pandas.DataFrame) and isinstance(args[1], pandas.DataFrame):
+      verb = cls(*args[1:], **kwargs)
+      return verb(args[0].copy(deep=True))
+    else:
+      return super(Verb, cls).__new__(cls)
 
-@ApplyToDataframe
-def right_join(right, **kwargs):
-  return mutating_join(right, how='right', **kwargs)
+    def __rrshift__(self, other):
+      return self.__call__(other)
+
+
+class inner_join(MutatingJoin):
+  """ Perform sql style inner join
+  """
+  __name__ = 'inner_join'
+
+  def __call__(self, df):
+    if self.kwargs:
```

Comment:
`self.kwargs` should always exist at this point-- it might be an empty `dict`, but it should exist. Why not pass just update it and pass the whole kwargs (removing the if-else)?